### PR TITLE
Request semaphore for mongo backend is now optional via CLI parameter

### PR DESCRIPTION
**Pull request by kzangeli**
Adding a new CLI parameter: -mutexPolicy (none/read/write/all) to turn off (all on is default) mutex protection for requests in mongoBackend.
The idea is to save some execution time, see Issue #910 for more info.

### CODE REVIEW

File: src/app/contextBroker/contextBroker.cpp [hunk 1]
```diff
@@ -211,6 +211,7 @@ bool            mtenant;
 char            rush[256];
 long            dbTimeout;
 long            httpTimeout;
+char            mutexPolicy[16];
```
Comment (fgalan) [inline src/app/contextBroker/contextBroker.cpp]: CHANGEX_NEXT_RELEASE entry about the new CLI parameter should be included.

File: src/lib/mongoBackend/mongoNotifyContext.cpp [hunk 2]
```diff
@@ -43,7 +43,9 @@ HttpStatusCode mongoNotifyContext
   const std::vector<std::string>&  servicePathV
 )
 {
-    reqSemTake(__FUNCTION__, "ngsi10 notification");
+    bool reqSemTaken;
+
+    reqSemTake(__FUNCTION__, "ngsi10 notification", SemReadOp, &reqSemTaken);
```
Comment (fgalan) [inline src/lib/mongoBackend/mongoNotifyContext.cpp]: I think mongoNotifyContext refers to the processing of notifyContextRequests. Thus, it is like a updateContext. In that case, SemWriteOp should be used.

File: src/lib/mongoBackend/mongoNotifyContextAvailability.cpp [hunk 3]
```diff
@@ -46,10 +46,11 @@ HttpStatusCode mongoNotifyContextAvailability
 )
 {
     const std::string notifyFormat = uriParam[URI_PARAM_NOTIFY_FORMAT];
+    bool              reqSemTaken;
 
     LM_T(LmtMongo, ("Notify Context Availability: '%s' format", notifyFormat.c_str()));
 
-    reqSemTake(__FUNCTION__, "mongo ngsi9 notification");
+    reqSemTake(__FUNCTION__, "mongo ngsi9 notification", SemReadOp, &reqSemTaken);
```
Comment (fgalan) [inline src/lib/mongoBackend/mongoNotifyContextAvailability.cpp]: I think mongoNotifyContextAvailabiltity refers to the processing of notifyContextRequests. Thus, it is like a registerContext. In that case, SemWriteOp should be used.

File: src/app/contextBroker/contextBroker.cpp (hunk 1 above)
Comment (kzangeli) [inline src/app/contextBroker/contextBroker.cpp]: Yes, now that develop is 'clean' I can add the line in CHANGES_NEXT_RELEASE

File: src/lib/common/sem.cpp [hunk 4]
```diff
@@ -83,14 +85,39 @@ int semInit(int shared, int takenInitially)
 *
 * reqSemTake -
 */
-int reqSemTake(const char* who, const char* what)
+int reqSemTake(const char* who, const char* what, SemRequestType reqType, bool* taken)
 {
   int r;
 
+  if (reqType == SemNoneOp)
```
Comment (fgalan) [inline src/lib/common/sem.cpp]: This is just a "theoretical" case, right? I mean, we don't have any SemNoneOp operation...

File: src/lib/mongoBackend/mongoNotifyContext.cpp (hunk 2 above)
Comment (kzangeli) [inline src/lib/mongoBackend/mongoNotifyContext.cpp]: ok, I was thinking about notifying, not about getting notified ...
If processContextElement modifies the database, it should definitely be SemWriteOp.

File: src/lib/mongoBackend/mongoNotifyContextAvailability.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/mongoBackend/mongoNotifyContextAvailability.cpp]: ok

File: src/lib/common/sem.cpp (hunk 4 above)
Comment (kzangeli) [inline src/lib/common/sem.cpp]: I didn't mean to ...
I guess I was thinking about reqPolicy and treated both of them the same way.
I will remove this part, thanks for discovering it :-)

File: src/app/contextBroker/contextBroker.cpp (hunk 1 above)
Comment (kzangeli) [inline src/app/contextBroker/contextBroker.cpp]: Fixed in 861eb29

File: src/lib/mongoBackend/mongoNotifyContext.cpp (hunk 2 above)
Comment (kzangeli) [inline src/lib/mongoBackend/mongoNotifyContext.cpp]: Fixed in 861eb29

File: src/lib/mongoBackend/mongoNotifyContextAvailability.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/mongoBackend/mongoNotifyContextAvailability.cpp]: Fixed in 861eb29

File: src/lib/common/sem.cpp (hunk 4 above)
Comment (kzangeli) [inline src/lib/common/sem.cpp]: Fixed in 861eb29

**Comment by fgalan** [comment]
LGTM

**Pull request merged**